workflow.go

```diff
@@ -0,0 +1,257 @@
+package subscription
+
+import (
+	"errors"
+	"time"
+
+	"go.temporal.io/sdk/workflow"
+)
+
+// This threshold is purposefully short to allow for easier
+// testing and demonstration of Continue-As-New. The Temporal Events in
+// this application are small enough that it's unlikely to hit the event
+// *size* limit, and so in the "real" version of this,
+// EventHistoryThreshold should be much closer to 50k.
+const EventHistoryThreshold = 50
+
+func SubscriptionWorkflow(ctx workflow.Context, subscription SubscriptionInfo) (err error) {
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Subscription workflow started", "customer", subscription.Customer, "subscription", subscription)
+
+	err = subscription.trialPeriod(ctx)
+	var continueAsNewErr *workflow.ContinueAsNewError
+	if err != nil && errors.As(err, &continueAsNewErr) {
```

Comment:
```suggestion
	if contErr, _ := err.(*workflow.ContinueAsNewError); contErr != nil {
```

No need to use the hierarchical `errors.As` here